### PR TITLE
fix: Resolve list relation regardless of order of protocol files.

### PR DESCRIPTION
**Pull request by Isakdl**
# Fix

Fixes a bug where the ordered of the protocol files mattered when resolving list relations.

Closes: https://github.com/serverpod/serverpod/issues/1397

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

### CODE REVIEW

File: tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart [hunk 1]
```diff
@@ -291,19 +291,37 @@ class EntityDependencyResolver {
     } else {
       var foreignFields = referenceClass.fields.where((field) {
         var fieldRelation = field.relation;
-        if (fieldRelation is! ForeignRelationDefinition) return false;
-        return fieldRelation.parentTable == classDefinition.tableName &&
-            fieldRelation.name == relation.name;
+        if (!(fieldRelation is UnresolvedObjectRelationDefinition ||
+            fieldRelation is ForeignRelationDefinition)) return false;
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart:295]: Suggestion: Place negation as close to check as possible.

Always hard to parse the impact on a negation on a group of operations. I think it is a bit simpler to be able to reach each check individually. (Extra line break after to separate the two returns.)

```suggestion
        if (fieldRelation is! UnresolvedObjectRelationDefinition &&
            fieldRelation is! ForeignRelationDefinition) {
             return false;
        }
        
```

File: tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart [hunk 2]
```diff
@@ -291,19 +291,37 @@ class EntityDependencyResolver {
...
+      if (foreignFields.isEmpty) return;
+
+      var foreignField = foreignFields.first;
+
+      String? foreignFieldName;
+
+      var foreignRelation = foreignField.relation;
+      if (foreignRelation is ForeignRelationDefinition) {
+        foreignFieldName = foreignField.name;
+      } else if (foreignRelation is UnresolvedObjectRelationDefinition) {
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart:308]: Question: For my knowledge, what has happened if we still have an unresolved object relation definition here?

File: tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart [hunk 3]
```diff
@@ -291,19 +291,37 @@ class EntityDependencyResolver {
...
+        foreignFieldName = foreignRelation.fieldName ??
+            _createImplicitForeignFieldName(foreignField.name);
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart:310]: Question: Why do we need to encode this into the name? Doesn't this need to be in sync with how we generate the code in the code generator now?

For the list we set the `implicitForeignField: true` so that the formatting can be handled in one place (the code generator). Would that be possible here as well?

File: tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_one_to_many_test.dart [hunk 4]
```diff
@@ -472,4 +472,113 @@ void main() {
...
+    test('then customer has list orders field', () {
+      expect(classDefinition.findField('orders'), isNotNull);
+    });
+
+    test('then orders is list relation', () {
+      expect(classDefinition.findField('orders')?.relation.runtimeType,
+          ListRelationDefinition);
+    });
+
+    test('then the list relation is NOT linked with the companyId field', () {
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_one_to_many_test.dart:577]:
```suggestion
    test('then the list relation is NOT linked with the customerId field', () {
```

File: tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_one_to_many_test.dart [hunk 5]
```diff
@@ -472,4 +472,113 @@ void main() {
...
+      var relation = classDefinition.findField('orders')?.relation
+          as ListRelationDefinition;
+
+      expect(relation.foreignFieldName, isNot('companyId'));
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_one_to_many_test.dart:581]:
```suggestion
      expect(relation.foreignFieldName, isNot('customerId'));
```

**Review by SandPod — COMMENTED** [review]
Good to get fixed! 🦖 

Couple of questions and suggestions.

File: tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart (hunk 3 above)
Comment (Isakdl) [reply tools/serverpod_cli/lib/src/analyzer/entities/entity_dependency_resolver.dart:310]: refactor name to: `_createImplicitForeignIdFieldName(...);`

File: tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_one_to_many_test.dart [hunk 6]
```diff
@@ -574,11 +574,11 @@ void main() {
           ListRelationDefinition);
     });
 
-    test('then the list relation is NOT linked with the companyId field', () {
+    test('then the list relation is NOT linked with the customerId field', () {
       var relation = classDefinition.findField('orders')?.relation
           as ListRelationDefinition;
 
-      expect(relation.foreignFieldName, isNot('companyId'));
+      expect(relation.foreignFieldName, isNot('customerId'));
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/entities/stateful_analyzer/protocol_validation/relation_one_to_many_test.dart:581]: Would be good to reflect this change on line 547 as well. Right now the field is called company in the protocol file.

**Review by SandPod — APPROVED** [review]
LGTM with the last comment.
